[core] Fix Regular Expression Denial of Service (ReDoS) vulnerabilities

SuperMaxine opened the pull request:
Fixes https://github.com/mui/material-ui/issues/44078

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/material-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

mj12albert commented on docs/src/modules/components/Demo.test.js:
```diff
@@ -0,0 +1,42 @@
+  }
+  str += '\n@.ts';
+
+  it('should complete within 10 seconds', () => {
```
Hey @SuperMaxine thanks for working on this, just wondering why `10 seconds`? At first glance it seems a bit arbitrary

SuperMaxine replied:
This is just a loose condition. In fact, if there is no ReDoS, the runtime of these tests in my local environment is around 0.0x seconds. However, if ReDoS exists (i.e., if my fixes in the source code are removed), these examples might hang for several minutes or even longer. Since I don’t have much information about the performance of the machines used for automated review checks, I chose an upper limit that would never be hit under normal situation. If convenient, please adjust this to a more practical value based on your actual needs.

mnajdova reviewed (commented):
A comment above the changed regular expressions would be enough, I wouldn't add tests specifically for a DOS attack. We could add a link to a example of how it can be tested.

SuperMaxine commented:
> A comment above the changed regular expressions would be enough, I wouldn't add tests specifically for a DOS attack. We could add a link to a example of how it can be tested.

Modified as described above, removed the tests and added comment descriptions, not sure how you want to add the link? I think the 4 PoCs in issue https://github.com/mui/material-ui/issues/44078 are good enough as examples.

[PoC_1.zip](https://github.com/user-attachments/files/17350200/PoC_1.zip)
[PoC_2.zip](https://github.com/user-attachments/files/17350201/PoC_2.zip)
[PoC_3.zip](https://github.com/user-attachments/files/17350203/PoC_3.zip)
[PoC_4.zip](https://github.com/user-attachments/files/17350202/PoC_4.zip)

mj12albert reviewed (approved):
@SuperMaxine I've also added a link to the issue as a comment as well, thanks again for working on this ~

oliviertassinari commented on docs/src/modules/components/Demo.js:
```diff
@@ -473,7 +473,9 @@ export default function Demo(props) {
         `The following demos use TS directly: ${demoOptions.demo}.`,
         '',
         'Please run "pnpm docs:typescript:formatted" to generate a JS version and reference it:',
-        `{{"demo": "${demoOptions.demo.replace(/\.(.*)$/, '.js')}", …}}.`,
```
I got curious, I looked closer at this. I see no catastrophic backtracking here https://www.regular-expressions.info/catastrophic.html but it's interesting how it's faster when there are a lot of dots:

before
<img width="560" alt="SCR-20241216-tgxa" src="https://github.com/user-attachments/assets/9116e78c-7f35-4e1e-94a1-8c82f3152260" />

after:
<img width="562" alt="SCR-20241216-tgyb" src="https://github.com/user-attachments/assets/d8b69129-52bb-4891-976c-6f1962c9e8f5" />

Now, in practice, this doesn't matter.

SuperMaxine replied:
Regex101 is suitable for learning regexes and testing the functionality of regexes in development. However, in serious software engineering and programming theory research, its behavior often deviates from the true regex engines behind the languages it lists. Additionally, from the image you provided, since it shows step numbers, I can infer that you might be using PHP's PCRE engine rather than JavaScript's regex engine in Regex101, which also has a certain degree of imprecision (perhaps I can try to provide a convenient external access to my instrumentation for migration counts of bytecode in real V8's regex engine, but it may take some time). For this regex, believe what you see; you can use a very simple example to confirm whether there is an efficiency issue:

```javascript
const str = "." + "a".repeat(100000) + "\n@.ts";
console.time('plain string');
const test1 = str.replace(/\.(.*)$/, '.js');
console.timeEnd('plain string');

const attackStr = "." + ".".repeat(100000) + "\n@.ts";
console.time('attack string');
const test2 = attackStr.replace(/\.(.*)$/, '.js');
console.timeEnd('attack string');
```

On my PC, it shows the following results:

```
plain string: 0.174072265625 ms
attack string: 2774.43994140625 ms
```

It easily demonstrates a performance gap of about 15,000 times. Preemptively making some adjustments to avoid this does not seem harmful as long as it does not affect actual functionality.

Regarding your statement that no catastrophic backtracking was found in this case, this example differs slightly from those on the web page you referenced; its backtracking behavior comes from JavaScript's default "substring matching" feature. That is to say, all matching behaviors of regex `<r>` in JavaScript can be approximately equivalent to matching `^.*<r>$` in engines like Java or Python. In this example, `.*\.(.*)` forms a pathological sub regex where the engine cannot distinguish whether each dot belongs to the first or second `.*`, leading to numerous attempts and backtracking. [Some studies](https://www.usenix.org/system/files/sec21-li-yeting.pdf) have categorized this as "Starting with Large Quantifier (SLQ)" pattern.

However, I see that this file is named `Demo.js`, so I understand why you think this issue would not cause much impact in your project. But if making a simple modification does not affect your functionality and can prevent others from potentially misusing or referencing code snippets unintentionally causing issues down the line, I still believe it benefits the software ecosystem overall without any harm done. Ultimately though, you are members of this project and I respect all decisions regarding how you handle your project.
